File: Haskell/Poker.hs
module Poker where
--Student Name: Mahamudul Islam
--Student Number: 500963051
--Student Name: Naushad Sayeed
--Student Number: 500589793
    import Control.Exception (assert)
    import Data.List (permutations, sort, sortOn, intersect, findIndices, (\\))
    import Data.Function (on)

    deal = trans
        . uncurry (winningHand 5)
        . hands
        . form where
            trans = sort . map (\n -> (show (([2..13] ++ [1]) !! (n % 13))) ++ [("CDHS" !! (n % 4))])
            hands (a:b:c:d:flop) = (a:c:flop, b:d:flop)
            indexOf x = head . findIndices (x==)
            encode (n:ns) [rank, suit]
                | n % 13 == indexOf rank "23456789TJQKA" && n % 4 == indexOf suit "CDHS" = n
                | otherwise = encode ns [rank, suit]
            form = map (\n -> encode [0..] [("A23456789TJQK" !! ((n - 1) % 13)), ("CDHS" !! ((n - 1) // 13))])

    winningHand n cards1 cards2
        | null cards1 = assert (null cards2) []
        | n < 1 = []
        | ranking1 > ranking2 = makeHand cards1 hand1
        | ranking1 < ranking2 = makeHand cards2 hand2
        | tieBroken == 1 = makeHand cards1 hand1
        | tieBroken == 2 = makeHand cards2 hand2
        | otherwise = hand1 ++ winningHand
                        (n - assert (length hand1 == length hand2) (length hand1))
                        (cards1 \\ hand1)
                        (cards2 \\ hand2)
        where
            makeHand cards hand = hand ++ (take (n - length hand) (sortOn ((13 -) . (%13)) (cards \\ hand))) 
            (ranking1, hand1) = bestHand cards1
            (ranking2, hand2) = bestHand cards2
            tieBroken = tieBreaker hand1 hand2

    bestHand hand = foldr1 (\xs ys -> if null (snd ys) then xs else ys) (map ($ hand) pokerHands)

    tieBreaker = tieBreaker' `on` highestVals where
        highestVals = reverse . arrangeVals . map (% 13) where
            arrangeVals vals = if vals `intersect` straightWithLowAce == straightWithLowAce
                then [-1..3]
                else sort vals
        tieBreaker' [] [] = 0
        tieBreaker' (card1:cards1) (card2:cards2)
            | card1 > card2 = 1
            | card1 < card2 = 2
            | otherwise = tieBreaker' cards1 cards2

    straightFlush = straight . flush
    fourOfAKind = eqOn 4 (% 13)
    fullHouse = pairAnd threeOfAKind
    flush = eqOn 5 (% 4)
    straight cards = (sort (map (% 13) cards) `elem` cons5) ? cards where
        cons5 = straightWithLowAce : [[n..n+4] | n <- [0..8]]
    threeOfAKind = eqOn 3 (% 13)
    twoPair = pairAnd pair
    pair = eqOn 2 (% 13)
    highCard = eqOn 1 id
    pokerHands = zipWith rank [1..] [
        highCard
        , pair
        , twoPair
        , threeOfAKind
        , straight
        , flush
        , fullHouse
        , fourOfAKind
        , straightFlush
        ] where
            rank ranking pokerHand cards = (ranking, handResult) where
                handResult = foldr1 bestVersionOfHand (map (pokerHand . take 5) $ permutations cards) where
                    bestVersionOfHand [] ys = ys
                    bestVersionOfHand xs [] = xs
                    bestVersionOfHand xs ys
                        | tieBreaker xs ys < 2 = xs
                        | otherwise = ys

    (%) = mod
    (//) = div
    straightWithLowAce = [0,1,2,3,12]
    tryTake n xs
        | length xs < n = []
        | otherwise = take n xs
    cond ? xs
        | cond = xs
        | otherwise = []
    eqOn n f cards = allEq (tryTake n $ map f cards) ? tryTake n cards
    allEq [] = False
    allEq (x:xs) = all (== x) xs
    pairAnd pokerHand cards
        | any null [picked1, picked2] = []
        | otherwise = picked1 ++ picked2
        where
            picked1 = pokerHand cards
            picked2 = pair $ cards \\ picked1

File: Poker.ex
defmodule Poker do
  import String, only: [capitalize: 1, codepoints: 1, to_atom: 1, to_integer: 1, at: 2, split: 2]
  import Enum, only: [reverse: 1, map: 2, join: 2, sort: 1, sort_by: 2]
  import List, only: [to_tuple: 1]
  import Tuple, only: [to_list: 1]

  def deal([a|[b|[c|[d|flop]]]]) do
    parse_input = &(&1 |> map(fn n -> at("A23456789TJQK", rem(n - 1, 13)) <> at("cdhs", div(n - 1, 13)) end) |> join(" "))
    format_output = &(&1 |> to_list |> map(fn {r, s} -> to_string(if card_value(r) == 14, do: 1, else: card_value(r)) <> capitalize(to_string(s)) end))
    hand1 = best_hand(parse_input.([a, c]), parse_input.(flop))
    hand2 = best_hand(parse_input.([b, d]), parse_input.(flop))
    format_output.(if hand_compare(hand1, hand2) == :gt, do: hand1, else: hand2) |> sort
  end

  defp best_hand(hand, flop) when is_binary(hand) do
    best_hand(parse_hand(hand), flop)
  end
  defp best_hand(hand, flop) when is_binary(flop) do
    best_hand(hand, parse_hand(flop))
  end
  defp best_hand(hand, flop) do
    comb(5, to_list(hand) ++ to_list(flop))
    |> sort_by(fn cards ->
      cards |> to_tuple |> hand_value
    end)
    |> reverse
    |> hd
    |> to_tuple
    |> sort_hand
  end

  defp comb(0, _), do: [[]]
  defp comb(_, []), do: []
  defp comb(m, [h|t]) do
    (for l <- comb(m-1, t), do: [h|l]) ++ comb(m, t)
  end

  defp hand_compare(hand1, hand2) when is_binary(hand1) do
    hand_compare(parse_hand(hand1), hand2)
  end
  defp hand_compare(hand1, hand2) when is_binary(hand2) do
    hand_compare(hand1, parse_hand(hand2))
  end
  defp hand_compare(hand1, hand2) do
    r = hand_value(hand1) - hand_value(hand2)

    cond do
      r > 0  -> :gt
      r == 0 -> :eq
      r < 0  -> :lt
    end
  end

  defp hand_value(str) when is_binary(str) do
    str |> parse_hand |> hand_value
  end
  defp hand_value(hand) do
    case hand_rank(hand) do
      {:straight_flush, a}        -> 8_000 + card_value(a)
      {:four_of_a_kind, _a, b}    -> 7_000 + card_value(b)
      {:full_house, a, b}         -> 6_000 + 15 * card_value(a) + card_value(b)
      {:flush, _r, a, b, c, d, e} -> 5_000 + card_value(a) + card_value(b) + card_value(c) + card_value(d) + card_value(e)
      {:straight, a}              -> 4_000 + card_value(a)
      {:three_of_a_kind, a, b, c} -> 3_000 + 15 * card_value(a) + card_value(b) + card_value(c)
      {:two_pair, a, b, c}        -> 2_000 + 15 * card_value(a) + 15 * card_value(b) + card_value(c)
      {:one_pair, a, b, c, d}     -> 1_000 + 15 * card_value(a) + card_value(b) + card_value(c) + card_value(d)
      {:high_card, a, b, c, d, e} -> card_value(a) + card_value(b) + card_value(c) + card_value(d) + card_value(e)
    end
  end

  defp hand_rank(str) when is_binary(str) do
    parse_hand(str) |> hand_rank
  end
  defp hand_rank(hand) do
    hand = sort_hand(hand)
    if is_straight(hand) do
      {{r1,_}, {r2,_}, _, _, _} = hand
      r = if r1 == :A && r2 == 5, do: 5, else: r1
      if is_flush(hand), do: {:straight_flush, r}, else: {:straight, r}
    else
      case hand do
        {{a,_},  {a,_},  {a,_},  {a,_},  {b,_}}  -> {:four_of_a_kind, a, b}
        {{b,_},  {a,_},  {a,_},  {a,_},  {a,_}}  -> {:four_of_a_kind, a, b}
        {{a,_},  {a,_},  {a,_},  {b,_},  {b,_}}  -> {:full_house, a, b}
        {{b,_},  {b,_},  {a,_},  {a,_},  {a,_}}  -> {:full_house, a, b}
        {{r1,a}, {r2,a}, {r3,a}, {r4,a}, {r5,a}} -> {:flush, a, r1, r2, r3, r4, r5}
        {{a,_},  {a,_},  {a,_},  {b,_},  {c,_}}  -> {:three_of_a_kind, a, b, c}
        {{b,_},  {a,_},  {a,_},  {a,_},  {c,_}}  -> {:three_of_a_kind, a, b, c}
        {{b,_},  {c,_},  {a,_},  {a,_},  {a,_}}  -> {:three_of_a_kind, a, b, c}
        {{a,_},  {a,_},  {b,_},  {b,_},  {c,_}}  -> {:two_pair, a, b, c}
        {{a,_},  {a,_},  {c,_},  {b,_},  {b,_}}  -> {:two_pair, a, b, c}
        {{c,_},  {a,_},  {a,_},  {b,_},  {b,_}}  -> {:two_pair, a, b, c}
        {{a,_},  {a,_},  {b,_},  {c,_},  {d,_}}  -> {:one_pair, a, b, c, d}
        {{b,_},  {a,_},  {a,_},  {c,_},  {d,_}}  -> {:one_pair, a, b, c, d}
        {{b,_},  {c,_},  {a,_},  {a,_},  {d,_}}  -> {:one_pair, a, b, c, d}
        {{b,_},  {c,_},  {d,_},  {a,_},  {a,_}}  -> {:one_pair, a, b, c, d}
        {{a,_},  {b,_},  {c,_},  {d,_},  {e,_}}  -> {:high_card, a, b, c, d, e}
      end
    end
  end

  defp is_straight(str) when is_binary(str) do
    str |> parse_hand |> is_straight
  end
  defp is_straight({{a,_}, {b,_}, {c,_}, {d,_}, {e,_}}) do
    (card_value(a) == card_value(b) + 1 || a == :A && b == 5) &&
      card_value(b) == card_value(c) + 1 &&
      card_value(c) == card_value(d) + 1 &&
      card_value(d) == card_value(e) + 1
  end

  defp is_flush({{_,a},{_,a},{_,a},{_,a},{_,a}}), do: true
  defp is_flush({_,_,_,_,_}),                     do: false

  defp card_value(:A), do: 14
  defp card_value(:K), do: 13
  defp card_value(:Q), do: 12
  defp card_value(:J), do: 11
  defp card_value(:T), do: 10
  defp card_value(i) when is_integer(i) and i >= 2 and i <= 9, do: i

  defp parse_hand(str) do
    str
    |> split(" ")
    |> map(&parse_card/1)
    |> to_tuple
  end

  defp parse_card(str) do
    [rank, suit] = codepoints(str)
    {parse_rank(rank), to_atom(suit)}
  end

  defp parse_rank("A"), do: :A
  defp parse_rank("K"), do: :K
  defp parse_rank("Q"), do: :Q
  defp parse_rank("J"), do: :J
  defp parse_rank("T"), do: :T
  defp parse_rank(str), do: to_integer(str)

  defp sort_hand(hand) do
    hand
    |> to_list
    |> sort_by(fn {rank,_} -> card_value(rank) end)
    |> reverse
    |> to_tuple
  end
end
